nano/node/websocket_payment_tracking.hpp

```diff
@@ -0,0 +1,109 @@
+	{
+	public:
+		payment_tracking_info (std::string id_a, std::shared_ptr<nano::state_block> const & tracked_block_a, nano::amount const & minimum_amount_a, nano::websocket::payment_tracker::policy tracking_policy_a, std::chrono::seconds track_until_a) :
+		id (id_a), tracked_block (tracked_block_a), minimum_amount (minimum_amount_a), last_sent_partial_amount (0), tracking_policy (tracking_policy_a), track_until (track_until_a)
+		{
+		}
+
+		/** The id provided through the Websocket subscription. This can be used by external systems to match up payment notifications. */
+		std::string id;
+
+		/** Tracked block, if any */
+		std::shared_ptr<nano::state_block> tracked_block;
+
+		/** The minimum amount required for a payment notification to be sent */
+		nano::amount minimum_amount;
+
+		/** If there's a partial payment (below minimum amount), we send a partial_payment notification. This is only done once per partial amount. */
+		nano::amount last_sent_partial_amount;
+
+		/** The requested tracking policy */
+		nano::websocket::payment_tracker::policy tracking_policy;
+
+		/** Tracking until this many seconds since epoch */
+		std::chrono::seconds track_until;
+	};
+
+	/** Start tracking a destination account based on websocket options for the payment subscription */
+	void track (nano::websocket::payment_tracking_options const & options_a);
+
+	/** Stop tracking this destination account */
+	void untrack (std::string const & account_a);
+
+	/**
+	 * Update the last-sent partial payment amount
+	 * @param account_a Account being tracked
+	 * @param amount_a The current total balance
+	 * @return true if the amount is different from what is previously recorded. This causes a partial_payment message to be sent.
+	 */
+	bool update_partial_payment_amount (std::string const & account_a, nano::amount const & amount_a);
+
+	/** Get tracking info for the given account, if available */
+	std::optional<payment_tracking_info> get_tracking_info (std::string const & account_a);
+
+	/** For each tracked account by this session, invoke \p callback_a */
+	void for_each (std::function<void(std::string const &, payment_tracking_info const &)> callback_a);
+
+private:
+	nano::locked<std::unordered_map<std::string, payment_tracking_info>> tracked_accounts;
+};
+
+/** Interacts with the node to hand off send blocks, and queries the ledger for confirmation status and balances */
+class payment_validator
+{
+public:
+	/** Starts ongoing payment tracking */
+	payment_validator (nano::node & node_a);
+
+	/** Check if the payment conditions are met, and if so, send notification to websocket client */
+	void check_payment (nano::account const & destination_account_a, nano::block_hash const & block_hash_a, std::shared_ptr<nano::websocket::session> const & session_a);
+
+	/** Publish a send state block to the network */
+	void publish_block (std::shared_ptr<nano::block> const & block_a, bool const work_watcher_a);
+
+private:
+	nano::node & node;
```

Comment:
It would be nice not to have to depend on the whole node for the payment validator (although we still do this in a lot of places). Most of the uses of the node look easy enough to remove and use the objects themselves (`ledger`, `logger` etc). The use of functions on the node aren't so easy, e.g `process_local`, probably could pass in a `publish_block` callback for that. It also makes it clearer what dependencies this class actually has. But it's quite a bit more effort, and I don't think there will be any initialization issues like we've had in other areas, so don't worry about it unless you feel inclined.

Reply:
Yeah exactly, tried to avoid it, but felt like process_local and friends should be factored out into something that can be passed along. I can take a look at a callback.

Reply:
Now passing in everything needed + callback for process_local.